src/catchup/CatchupManagerImpl.cpp

```diff
@@ -186,35 +197,70 @@ CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
                                    "ledger {:d} was closed on network"),
                         lastLedgerInBuffer);
 
-        startOnlineCatchup();
+        startOnlineCatchup(false);
```

Comment:
The main issue is that we're not sending LedgerCloseData for the checkpoint ledger, but an SCP message. When the out of sync node receives the SCP message, it does not trigger `processLedger` because it does not receive a ledger that needs to be buffered. The special case is required for early catchup so that catchup knows to check the SCP messages for the relevant hash instead of checking the buffered ledgers. This might be bad design on my part, should we be sending LedgerCloseData instead of SCP messages for the checkpoint ledger? I also might be missing something here because I'm still not well versed on how SCP messages become LedgerCloseData to be processed and buffered.

Reply:
Oh so a bit of background on how nodes externalize. Sending SCP messages is the right thing to do here. When enough SCP messages to reach consensus are received, and processed by the SCP state machine, we can officially "externalize" that ledger. Externalize means your quorum reached consensus, but you might not necessarily be in the right state to apply that ledger (e.g. last applied ledger is N, but N+5 was externalized). This information is passed down to Herder and LedgerManager, which tries to either apply that ledger or start catchup to it. See `HerderImpl::processExternalized`, it constructs `LedgerCloseData`, which is then passed to catchup manager to potentially trigger catchup. So we should be able to buffer that ledger just like any other ledger. 

I'd check that the checkpoint ledger we receive is actually externalized by Herder and passed down to CatchupManager. One other thing to note here is that we trim buffered ledgers once in a while in CatchupManager (see `CatchupManagerImpl::trimSyncingLedgers`). I don't think it should affect the desired buffering logic here, but is it possible that the test isn't setup correctly such that the checkpoint ledger is trimmed?